hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvocationMonitor.java

```diff
@@ -47,134 +61,178 @@
+        scheduler.scheduleAtFixedRate(new BroadcastOperationHeartbeatsTask(), 0, heartbeatBroadcastPeriodMillis, MILLISECONDS);
+        return scheduler;
+    }
+
+    private long invocationTimeoutMillis() {
+        GroupProperties props = nodeEngine.getGroupProperties();
+        long heartbeatTimeoutMillis = props.getMillis(OPERATION_CALL_TIMEOUT_MILLIS);
 
-        metricsRegistry.scanAndRegister(this, "operation.invocations");
+        if (logger.isFinestEnabled()) {
+            logger.finest("Operation invocation timeout is " + heartbeatTimeoutMillis + " ms");
+        }
+
+        return heartbeatTimeoutMillis;
+    }
+
+    private long heartbeatBroadcastPeriodMillis() {
+        // The heartbeat period is configured to be 1/4 of the call timeout. So with default settings, every 15 seconds,
+        // every member in the cluster, will notify every other member in the cluster about all calls that are pending.
+        // This is done quite efficiently; imagine at any given moment one node has 1000 concurrent calls pending on another
+        // node; then every 15 seconds an 8 KByte packet is send to the other member.
+        int callTimeoutMs = nodeEngine.getGroupProperties().getInteger(OPERATION_CALL_TIMEOUT_MILLIS);
+        long periodMs = Math.max(SECONDS.toMillis(1), callTimeoutMs / HEARTBEAT_CALL_TIMEOUT_RATIO);
+
+        if (logger.isFinestEnabled()) {
+            logger.finest("Operation heartbeat period is " + periodMs + " ms");
+        }
 
-        monitorThread.start();
+        return periodMs;
+    }
+
+    private long slowInvocationThresholdMillis() {
+        GroupProperties props = nodeEngine.getGroupProperties();
+        long thresholdMs = props.getMillis(SLOW_INVOCATION_DETECTOR_THRESHOLD_MILLIS);
+        if (thresholdMs > -1) {
+            logger.info("Slow invocation detector enabled, using threshold: " + thresholdMs + " ms");
+        }
+        return thresholdMs;
     }
 
     public void shutdown() {
-        monitorThread.shutdown();
+        scheduler.shutdown();
     }
 
     public void awaitTermination(long timeoutMillis) throws InterruptedException {
-        monitorThread.join(timeoutMillis);
+        scheduler.awaitTermination(timeoutMillis, MILLISECONDS);
     }
 
     public void onMemberLeft(MemberImpl member) {
         // postpone notifying invocations since real response may arrive in the mean time.
-        Runnable task = new OnMemberLeftTask(member);
-        executionService.schedule(task, ON_MEMBER_LEFT_DELAY_MS, TimeUnit.MILLISECONDS);
+        scheduler.schedule(new OnMemberLeftTask(member), ON_MEMBER_LEFT_DELAY_MILLIS, MILLISECONDS);
+    }
+
+    @Override
+    public void handle(Packet packet) {
+        scheduler.execute(new ProcessOperationHeartbeatsTask(packet));
     }
 
     /**
-     * The InvocationMonitorThread iterates over all pending invocations and sees what needs to be done
+     * The MonitorTask iterates over all pending invocations and sees what needs to be done
```

Comment:
could you clarify the contract? "what needs to be done" is a very vague responsibility. 
![image](https://cloud.githubusercontent.com/assets/158619/14021568/b0b5104e-f1e4-11e5-95a9-f783f34b5ec0.png)